cloudevents/tests/test_pydantic_cloudevent.py

```diff
@@ -0,0 +1,347 @@
+    dummy_attributes, json_content_type
+):
+    dummy_attributes["datacontenttype"] = json_content_type
+    assert loads(CloudEvent(dummy_attributes, data='{"hello": "world"}',).json())[
+        "data"
+    ] == {"hello": "world"}
+
+
+@pytest.fixture(
+    params=[
+        pytest.param("video/mp2t", id="MPEG transport stream"),
+        pytest.param("text/plain", id="Text, (generally ASCII or ISO 8859-n)"),
+        pytest.param("application/vnd.visio", id="Microsoft Visio"),
+        pytest.param("audio/wav", id="Waveform Audio Format"),
+        pytest.param("audio/webm", id="WEBM audio"),
+        pytest.param("video/webm", id="WEBM video"),
+        pytest.param("image/webp", id="WEBP image"),
+        pytest.param("application/gzip", id="GZip Compressed Archive"),
+        pytest.param("image/gif", id="Graphics Interchange Format (GIF)"),
+        pytest.param("text/html", id="HyperText Markup Language (HTML)"),
+        pytest.param("image/vnd.microsoft.icon", id="Icon format"),
+        pytest.param("text/calendar", id="iCalendar format"),
+        pytest.param("application/java-archive", id="Java Archive (JAR)"),
+        pytest.param("image/jpeg", id="JPEG images"),
+    ]
+)
+def non_json_type_dummy_attributes(dummy_attributes, request):
+    dummy_attributes["datacontenttype"] = request.param
+    return dummy_attributes
+
+
+def test_json_data_serialization_with_explicit_non_json_content_type(
+    non_json_type_dummy_attributes,
+):
+    assert (
+        loads(
+            CloudEvent(
+                non_json_type_dummy_attributes,
+                data='{"hello": "world"}',
+            ).json()
+        )["data"]
+        == '{"hello": "world"}'
+    )
+
+
+def test_binary_data_serialization(non_json_type_dummy_attributes):
+    result_json = loads(
+        CloudEvent(
+            non_json_type_dummy_attributes,
+            data=b"\x00\x00\x11Hello World",
+        ).json()
+    )
+    assert result_json["data_base64"] == "AAARSGVsbG8gV29ybGQ="
+    assert "daata" not in result_json
+
+
+def test_binary_data_deserialization():
+    assert CloudEvent.parse_raw(
+        b'{"source": "dummy:source", "id": "11775cb2-fd00-4487-a18b-30c3600eaa5f",'
+        b' "type": "dummy.type", "specversion": "1.0", "time":'
+        b' "2022-07-16T12:03:20.519216+00:00", "subject": null, "datacontenttype":'
+        b' "application/octet-stream", "dataschema": null, "data_base64":'
+        b' "AAARSGVsbG8gV29ybGQ="}'
+    ).dict() == {
+        "data": b"\x00\x00\x11Hello World",
+        "datacontenttype": "application/octet-stream",
+        "dataschema": None,
+        "id": "11775cb2-fd00-4487-a18b-30c3600eaa5f",
+        "source": "dummy:source",
+        "specversion": SpecVersion.v1_0,
+        "subject": None,
+        "time": datetime.datetime(
+            2022, 7, 16, 12, 3, 20, 519216, tzinfo=datetime.timezone.utc
+        ),
+        "type": "dummy.type",
+    }
```

Comment:
That's usually more readable if you extract an `actual` and `expected` variables.